File: data/src/main/java/com/clloret/days/data/remote/entities/ApiEvent.java
package com.clloret.days.data.remote.entities;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.clloret.days.domain.entities.Event.TimeUnit;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import java.util.Date;

public class ApiEvent {

  private static final String[] EMPTY_ARRAY = new String[0];

  @Expose(serialize = false)
  private transient String id;

  @Expose(serialize = false)
  private transient String createdTime;

  @SerializedName("Name")
  private String name;

  @SerializedName("Description")
  private String description;

  @SerializedName("Date")
  private Date date;

  @SerializedName("Tags")
  private String[] tags = EMPTY_ARRAY;

  @SerializedName("Favorite")
  private boolean favorite;

  @SerializedName("Reminder")
  private Integer reminder;

  @SerializedName("ReminderUnit")
  private String reminderUnit;

  @SerializedName("TimeLapse")
  private int timeLapse;

  @SerializedName("TimeLapseUnit")
  private String timeLapseUnit;

  // empty constructor needed by the Airtable library
  @SuppressWarnings("unused")
  public ApiEvent() {

  }

  public ApiEvent(@Nullable String id) {

    this.id = id;
  }

  public String getId() {

    return id;
  }

  public void setId(String id) {

    this.id = id;
  }

  public String getName() {

    return name;
  }

  public void setName(String name) {

    this.name = name;
  }

  public Date getDate() {

    return (Date) date.clone();
  }

  public void setDate(Date date) {

    this.date = (Date) date.clone();
  }

  public boolean isFavorite() {

    return favorite;
  }

  public void setFavorite(boolean favorite) {

    this.favorite = favorite;
  }

  public String getCreatedTime() {

    return createdTime;
  }

  public void setCreatedTime(String createdTime) {

    this.createdTime = createdTime;
  }

  public String[] getTags() {

    return tags;
  }

  public void setTags(String[] tags) {

    this.tags = tags;
  }

  public String getDescription() {

    return description;
  }

  public void setDescription(String description) {

    this.description = description;
  }

  public Integer getReminder() {

    return reminder;
  }

  public void setReminder(Integer reminder) {

    this.reminder = reminder;
  }

  public String getReminderUnit() {

    return reminderUnit;
  }

  public void setReminderUnit(@NonNull String reminderUnit) {

    this.reminderUnit = reminderUnit;
  }

  public int getTimeLapse() {

    return timeLapse;
  }

  public void setTimeLapse(int timeLapse) {

    this.timeLapse = timeLapse;
  }

  public String getTimeLapseUnit() {

    return timeLapseUnit;
  }

  public void setTimeLapseUnit(String timeLapseUnit) {

    this.timeLapseUnit = timeLapseUnit;
  }

  public TimeUnit getReminderTimeUnit() {

    return reminderUnit == null ? null : TimeUnit.valueOf(reminderUnit.toUpperCase());
  }

  public void setReminderTimeUnit(TimeUnit reminderUnit) {

    this.reminderUnit = reminderUnit == null ? null : reminderUnit.toString();
  }

  public TimeUnit getTimeLapseTimeUnit() {

    return timeLapseUnit == null ? null : TimeUnit.valueOf(timeLapseUnit.toUpperCase());
  }

  public void setTimeLapseTimeUnit(TimeUnit timeLapseUnit) {

    this.timeLapseUnit = timeLapseUnit == null ? null : timeLapseUnit.toString();
  }

}

File: data/src/main/java/com/clloret/days/data/remote/entities/ApiEvent.kt
package com.clloret.days.data.remote.entities

import com.clloret.days.domain.entities.Event
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import java.util.*

@Suppress("unused")
class ApiEvent {
  @Expose(serialize = false)
  @Transient
  var id: String? = null

  @Expose(serialize = false)
  @Transient
  var createdTime: String? = null

  @SerializedName("Name")
  var name: String = ""

  @SerializedName("Description")
  var description: String? = null

  @SerializedName("Date")
  private var date: Date? = null

  @SerializedName("Tags")
  var tags: Array<String> = EMPTY_ARRAY

  @SerializedName("Favorite")
  var favorite = false

  @SerializedName("Reminder")
  var reminder: Int? = null

  @SerializedName("ReminderUnit")
  var reminderUnit: String? = null
    private set

  @SerializedName("TimeLapse")
  var timeLapse = 0

  @SerializedName("TimeLapseUnit")
  var timeLapseUnit: String? = null

  @SerializedName("ProgressDate")
  var progressDate: Date? = null

  // empty constructor needed by the Airtable library
  constructor()
  constructor(id: String?) {
    this.id = id
  }

  constructor(id: String?,
              name: String,
              description: String?,
              date: Date?,
              tags: Array<String>,
              favorite: Boolean,
              reminder: Int?,
              reminderTimeUnit: Event.TimeUnit,
              timeLapse: Int,
              timeLapseTimeUnit: Event.TimeUnit,
              progressDate: Date?) {
    this.id = id
    this.name = name
    this.description = description
    this.date = date
    this.tags = tags
    this.favorite = favorite
    this.reminder = reminder
    this.reminderTimeUnit = reminderTimeUnit
    this.timeLapse = timeLapse
    this.timeLapseTimeUnit = timeLapseTimeUnit
    this.progressDate = progressDate
  }

  fun getDate(): Date {
    return date!!.clone() as Date
  }

  fun setDate(date: Date?) {
    this.date = date?.clone() as Date
  }

  fun setReminderUnit(reminderUnit: String) {
    this.reminderUnit = reminderUnit
  }

  var reminderTimeUnit: Event.TimeUnit
    get() = if (reminderUnit == null) Event.TimeUnit.DAY else
      Event.TimeUnit.valueOf(reminderUnit!!.toUpperCase(Locale.ROOT))
    set(reminderUnit) {
      this.reminderUnit = reminderUnit.toString()
    }

  var timeLapseTimeUnit: Event.TimeUnit
    get() = if (timeLapseUnit == null) Event.TimeUnit.DAY else
      Event.TimeUnit.valueOf(timeLapseUnit!!.toUpperCase(Locale.ROOT))
    set(timeLapseUnit) {
      this.timeLapseUnit = timeLapseUnit.toString()
    }

  companion object {
    private val EMPTY_ARRAY = emptyArray<String>()
  }
}